pallets/fiat_filter/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+        <T as frame_system::Config>::BlockNumber,
+    {
+        /// on set_dock_usd_rate executed
+        /// event parameters: [new_dock_usd_rate]
+        DockFiatRateUpdated(Permill),
+        /// on root_set_update_freq executed
+        /// event parameters: [new_update_frequency_blocks]
+        UpdateFreqUpdated(BlockNumber),
+    }
+}
+
+// Errors inform users that something went wrong.
+decl_error! {
+    pub enum Error for Module<T: Config> {
+        /// Failed calculation because of numeric type overflow
+        ArithmeticOverflow,
+        /// checked_div: tried to divide by zero
+        DivideByZero,
+        /// Call is not among the core_mods ones that should go through fiat_filter
+        UnexpectedCall,
+    }
+}
+
+// The module's callable functions (Dispatchable functions)
+// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+decl_module! {
+    pub struct Module<T: Config> for enum Call where origin: T::Origin {
+        // Errors must be initialized if they are used by the pallet.
+        type Error = Error<T>;
+        // Events must be initialized if they are used by the pallet.
+        fn deposit_event() = default;
+        /// `on_initialize` gets called on every new block
+        fn on_initialize(n: T::BlockNumber) -> Weight {
+            // update DockUsdRate if more than N blocks
+            let current_block = <system::Module<T>>::block_number();
+            if current_block - Self::last_updated_at() > Self::update_freq() {
+                // if let Err(e) = T::UpdaterDockFiatRate::update_dock_fiat_rate() {
+                //     sp_runtime::print(e);
+                // }
+                use common::PriceProvider;
+                if let Some((rate_per1000,_weight)) = T::PriceProvider::optimized_get_dock_usd_price() {
+                    // the only implementation of PriceProvider returns a PerThousand value (as a u32)
+                    let rate_permill = Permill::from_perthousand(rate_per1000);
+                    Self::set_dock_fiat_rate_(rate_permill)
+                }
+            };
+            return 0;
+        }
+
+
+
+        /// Execute a Call. Must be a DID or Blob call
+        #[weight = (10_000 + T::DbWeight::get().writes(1), Pays::No)]
+        pub fn execute_call(origin, call: Box<<T as Config>::Call>) -> DispatchResultWithPostInfo {
+            Self::execute_call_(origin, &call)?;
+            Ok(Pays::No.into())
+        }
+
+        /// Set the Update Frequency through Root
+        #[weight = 10_000 + T::DbWeight::get().writes(1)]
+        pub fn root_set_update_freq(origin, new_update_freq: T::BlockNumber) -> DispatchResult {
+            ensure_root(origin)?;
+            Self::set_update_freq_(new_update_freq);
+            Ok(())
+        }
+
+        /// Set DockFiatRate through Root
+        #[weight = 10_000 + T::DbWeight::get().writes(1)]
+        pub fn root_set_dock_fiat_rate(origin, new_rate: Permill) -> DispatchResult {
```

Comment:
If we want to avoid checking for price periodically but call `PriceProvider`'s every time we compute the fees while still having the possibility of root overriding the price, we can have `DockFiatRate` as an `Option` and when set (not `None`), it won't call `PriceProvider` but if unset, it will use `PriceProvider`